test/tbb/test_collaborative_call_once.cpp

```diff
@@ -0,0 +1,372 @@
+TEST_CASE("handles exceptions - state reset") {
+
+    std::atomic<bool> b{ false };
+    auto setB = [&b]() { b = true; };
+    auto setBAndCancel = [&b]() {
+        b = true;
+        throw call_once_exception{};
+    };
+
+    tbb::collaborative_once_flag flag;
+    REQUIRE_THROWS_AS(tbb::collaborative_call_once(flag, setBAndCancel), call_once_exception);
+    REQUIRE(b);
+
+    b = false;
+    REQUIRE_THROWS_AS(tbb::collaborative_call_once(flag, setBAndCancel), call_once_exception);
+    REQUIRE(b);
+
+    b = false;
+    tbb::collaborative_call_once(flag, setB);
+    REQUIRE(b);
+
+    b = false;
+    tbb::collaborative_call_once(flag, setB); // Now the call_once flag should be set.
+    REQUIRE(!b);
+
+    b = false;
+    REQUIRE_NOTHROW(tbb::collaborative_call_once(flag, setBAndCancel)); // Flag still set, so it shouldn't be called.
+    REQUIRE(!b);
+}
+
+//! Stress test for collaborative_call_once exception handling
+//! \brief \ref error_guessing \ref stress
+TEST_CASE("handles exceptions - stress test") {
+    std::atomic<int> data{0};
+
+    bool run_again{true};
+
+    auto throwing_func = [&] {
+        utils::doDummyWork(10000);
+        if (data < 100) {
+            data++;
+            run_again = true;
+            throw call_once_exception{};
+        }
+        run_again = false;
+    };
+
+    tbb::collaborative_once_flag flag;
+
+    utils::NativeParallelFor(utils::get_platform_max_threads()*5, [&](std::size_t) {
+        while(run_again) {
+            try {
+                tbb::collaborative_call_once(flag, throwing_func);
+            } catch(const call_once_exception&) {
+            }
+        }
+    });
+    REQUIRE(data == 100);
+}
+
+#endif
+
+//! Test for multiple help from moonlighting threads
+//! \brief \ref interface \ref requirement
+TEST_CASE("multiple help") {
+    // Test multiple threads do the work:
+    std::mutex mutex;
+    std::map<int, int> id_ct_map;
+    std::atomic<int> call_ct{0};
+
+    const size_t num_work_pieces = tbb::this_task_arena::max_concurrency() * 5;
+    // const auto time_per_work_piece = std::chrono::microseconds(5);
+    // // this allows us to make sure all the calls in the outer loop get stuck (if they will) before the inner loop begins
+    // const auto time_initial_delay = std::chrono::milliseconds(100);
+
+    // auto busy_wait = [](std::chrono::microseconds time) {
+    //     auto start_t = std::chrono::steady_clock::now();
+    //     while (std::chrono::steady_clock::now() < start_t + time) {
+    //         // busy wait
+    //     }
+    // };
+    auto work = [&]() {
+        call_ct++;
+        // busy_wait(time_initial_delay);
+        utils::doDummyWork(100000);
+        tbb::parallel_for(tbb::blocked_range<size_t>(0, num_work_pieces, 1), [&](const tbb::blocked_range<size_t>& range) {
+            for (size_t i = range.begin(); i != range.end(); ++i) {
+                // busy_wait(time_per_work_piece);
+                utils::doDummyWork(100000);
```

Comment:
It seems as synchronization based on sleeps. Consider using barrier to guarantee that all threads joined the parallel loop.